### PR TITLE
[#35] Implement Unit Tests for 'parseOwnerRepo' function

### PR DESCRIPTION
**Pull request by kahlil29**
- [x] Implemented 4 Unit tests as per the 4 formats that the function is supposed to work for.
- [x] Formatted the `test/Main.hs` file with `stylish-haskell`

Do let me know if anything needs to be changed/corrected/added 😄

### CODE REVIEW

File: test/Main.hs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+module Main where
+
+import GitHub.Data.Name
+import Hit.Issue (parseOwnerRepo)
+import Test.Hspec
+
+
+main :: IO ()
+main = hspec $ do
+  describe "parseOwnerRepo" $ do
```
Comment (chshersh) [inline test/Main.hs]: According to our style guide everything should be indented with 4 spaces

File: test/Main.hs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main where
+
+import GitHub.Data.Name
+import Hit.Issue (parseOwnerRepo)
+import Test.Hspec
```
Comment (chshersh) [inline test/Main.hs]: Could you please make all imports to use explicit import lists?

File: test/Main.hs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main where
+
+import GitHub.Data.Name
+import Hit.Issue (parseOwnerRepo)
```
Comment (chshersh) [inline test/Main.hs]: According to our style guide, you should group imports into multiple groups and put a blank line between groups. Groups are:

1. Dependencies from Hackage
2. Dependencies from the project itself

File: hit-on.cabal [hunk 4]
```diff
@@ -81,4 +81,20 @@ executable hit
   ghc-options:         -Wall
                        -threaded
                        -rtsopts
-                       -with-rtsopts=-N
\ No newline at end of file
+                       -with-rtsopts=-N
+
+
+test-suite hit-on-test
+  type:                exitcode-stdio-1.0
```
Comment (chshersh) [inline hit-on.cabal]: We have `common-options` stanza. You can import it as a first line of your `test-suite` stanza :slightly_smiling_face: 

After doing this you can remove some redundant code:

1. `-Wall` from `ghc-options`
2. `LANGUAGE` pragmas from your `Main.hs` file (they are already included into `common-options`)

File: hit-on.cabal [hunk 5]
```diff
@@ -81,4 +81,20 @@ executable hit
...
+  hs-source-dirs:      test
+  main-is:             Main.hs
+  build-depends:       base
+                     , text
+                     , hspec
+                     , github
+                     , hit-on
+  other-modules:
```
Comment (chshersh) [inline hit-on.cabal]: This section is empty, so you can remove it. Later, when we will have many tests, we will, of course, add new modules.

File: test/Main.hs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import Test.Hspec
+
+
+main :: IO ()
+main = hspec $ do
+  describe "parseOwnerRepo" $ do
+    let expectedOwnerName = N "kowainik"
+    let expectedRepoName = N "hit-on"
+    it "Parses a GitHub repo link (SSH) and gives use the Owner Name and Repo Name separately" $ do
+      parseOwnerRepo "git@github.com:kowainik/hit-on.git" `shouldBe` (Just (expectedOwnerName, expectedRepoName))
```
Comment (chshersh) [inline test/Main.hs]: I propose to move `Just (...)` into a separate variable like `expectedCredentials`. Just to reduce code duplication a bit.

File: test/Main.hs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    it "Parses a GitHub repo link (SSH) and gives use the Owner Name and Repo Name separately" $ do
+      parseOwnerRepo "git@github.com:kowainik/hit-on" `shouldBe` (Just (expectedOwnerName, expectedRepoName))
+
+
```
Comment (chshersh) [inline test/Main.hs]: I think two empty lines between test cases is not necessary

File: test/Main.hs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    it "Parses a GitHub repo link (HTTPS) and gives use the Owner Name and Repo Name separately" $ do
+      parseOwnerRepo "https://github.com/kowainik/hit-on.git" `shouldBe` (Just (expectedOwnerName, expectedRepoName))
+
+
+    it "Parses a GitHub repo link (HTTPS) and gives use the Owner Name and Repo Name separately" $ do
```
Comment (chshersh) [inline test/Main.hs]: You've tested _happy paths_ -- when function works. But usually, tests should include cases where a function can't return value. In our case you can add couple tests where `parseOwnerRepo` returns `Nothing`. Just to increase test coverage.

File: test/Main.hs (hunk 3 above)
Comment (kahlil29) [reply test/Main.hs]: Didn't know about this. Thanks for pointing it out!

File: hit-on.cabal (hunk 4 above)
Comment (kahlil29) [reply hit-on.cabal]: This makes a lot of sense. Making the change!

File: test/Main.hs (hunk 8 above)
Comment (kahlil29) [reply test/Main.hs]: This is an important thing, which I missed out... I'll include these things. Thanks a lot for the detailed feedback!

File: hit-on.cabal (hunk 4 above)
Comment (kahlil29) [reply hit-on.cabal]: Due to the import of `common-options`, I got a warning due to `[-Wmissing-export-lists]` so I added the `main` function to the export list at the start of the module (module name declaration).

File: test/Main.hs (hunk 8 above)
Comment (kahlil29) [reply test/Main.hs]: I've added 4 more cases of incorrect/broken links that are expected to return `Nothing`. 
Do let me know if you can think of more test cases. I'll add them! 😄

**Comment by kahlil29** [comment]
@chshersh Thanks for taking the time to review this PR in detail. 
Even though it's a small PR, there were quite a few things(formatting, etc) I missed out. 
Since I'm not used to raising PRs often in `kowainik`, I am not extremely familiar with the style guide and I hope that in my next PR, these errors would be less/not present 👍 

I have made the requested changes and pushed them. CI seems to have passed successfully!

Do let me know if there's anything else to be done before it can be merged into `master` 😄

**Pull request merged**